clang/test/Analysis/Checkers/CHERI/provenance-source.c

```diff
@@ -0,0 +1,159 @@
+  intptr_t a = d;
+  intptr_t b = (intptr_t)p;
+
+  intptr_t s = b + a; // expected-warning{{Ambiguous provenance: LHS is pointer, RHS is integer}}
+                      // expected-warning@-1{{binary expression on capability types 'intptr_t' (aka '__intcap') and 'intptr_t'; it is not clear which should be used as the source of provenance; currently provenance is inherited from the left-hand side}}
+  return *(char*)s;// expected-warning{{intcap with ambiguous provenance is used as capability}}
+}
+
+char right_prov(int d, char *p) {
+  intptr_t a = d;
+  intptr_t b = (intptr_t)p;
+
+  intptr_t s = a + b; // expected-warning{{Ambiguous provenance: LHS is integer, RHS is pointer}}
+                      // expected-warning@-1{{binary expression on capability types 'intptr_t' (aka '__intcap') and 'intptr_t'; it is not clear which should be used as the source of provenance; currently provenance is inherited from the left-hand side}}
+  return *(char*)s;// expected-warning{{intcap with ambiguous provenance is used as capability}}
+}
+
+char both_prov(int* d, char *p) {
+  intptr_t a = (intptr_t)d;
+  intptr_t b = (intptr_t)p;
+
+  intptr_t s = a + b; // expected-warning{{Ambiguous provenance: LHS and RHS are pointers}}
+                      // expected-warning@-1{{binary expression on capability types 'intptr_t' (aka '__intcap') and 'intptr_t'; it is not clear which should be used as the source of provenance; currently provenance is inherited from the left-hand side}}
+  return *(char*)s; // expected-warning{{intcap with ambiguous provenance is used as capability}}
+}
+
+char no_prov(int d, char p) {
+  intptr_t a = (intptr_t)d;
+  intptr_t b = (intptr_t)p;
+
+  intptr_t s = a + b; // expected-warning{{Ambiguous provenance: LHS and RHS are not pointers}}
+                      // expected-warning@-1{{binary expression on capability types 'intptr_t' (aka '__intcap') and 'intptr_t'; it is not clear which should be used as the source of provenance; currently provenance is inherited from the left-hand side}}
+  return *(char*)s; // expected-warning{{NULL-derived capability is used as pointer}}
+}
+
+char right_prov_cond(int d, char *p, int x) {
+  intptr_t a = d;
+  intptr_t b;
+  if (d > 42)
+    b = (intptr_t)p;
+  else
+    b = x;
+
+  intptr_t s = a + b; // expected-warning{{Ambiguous provenance: LHS is integer, RHS is pointer}}
+                      // expected-warning@-1{{Ambiguous provenance: LHS and RHS are not pointers}}
+                      // expected-warning@-2{{binary expression on capability types 'intptr_t' (aka '__intcap') and 'intptr_t'; it is not clear which should be used as the source of provenance; currently provenance is inherited from the left-hand side}}
+  return *(char*)s;// expected-warning{{intcap with ambiguous provenance is used as capability}}
+                   // expected-warning@-1{{NULL-derived capability is used as pointer}}
+}
+
+intptr_t no_bug(int *d, int *p) {
+  intptr_t a = (intptr_t)d;
```

Comment:
Please make sure tests are free of dead stores otherwise changes to the default set of warnings will break them